File: data-structures/linked_list/link_list.py
import json
#from ll_merge import mergeList


# **********************************
class LinkNode():

    value = ''
    next = None
    prev = None

    def __init__(self, value, next=None, prev=None):
        # constructor
        self.value = value
        self.next = next
        self.prev = prev


# **********************************
class LinkList():

    head = None

    def __init__(self):
        # constructor
        self.head = None

    def toJSON(self):
        # dump object to JSON and return as String
        buf = json.dumps(self, default=lambda o: o.__dict__, indent=0, separators=(',', ': '))
        buf = buf.replace('\n', ' ').replace('\r', '')
        return buf

    def toStr(self):
        # dump list as simple text-list

        # c,cnt are limiters to make sure we don't go run away
        # yes, we need cnt=self.count() and not -1, as we walk off list
        c, cnt = 0, self.count()
        buf = ''
        ptr = self.head
        while ptr is not None:
            buf = buf + ptr.value + ','
            ptr = ptr.next
            c += 1
            if (c > cnt):
                raise AssertionError(f'WAIT!!! Forever Loop!\nRecursive LinkList/Node\nbuf:[{buf}]')

        if buf.endswith(','):
            buf = buf[:-1]
        return buf

    def insert(self, value):
        # insert value at the head of the list
        node = LinkNode(value, self.head)
        node.next = self.head
        self.head = node

    def includes(self, value):
        # traverse list and determine if a value exists
        # return bool
        ret = False
        cur = self.head
        while cur is not None:
            if cur.value == value:
                ret = True
                break
            cur = cur.next
        return ret

    def count(self):
        # count the number of nodes and return count
        cnt = 0
        cur = self.head
        while cur is not None:
            cnt += 1
            cur = cur.next
        return cnt

    def append(self, value):
        # adds a new node with the given value to the end of the list
        # BigO == O(n)

        # walk to end of list
        prev, cur = None, self.head
        while cur is not None:
            prev = cur
            cur = cur.next

        # create the node and add it to the end
        node = LinkNode(value, None)
        if (prev == None):
            self.head = node
        else:
            prev.next = node

        return True

    def insertBefore(self, targetVal: int, newVal: str, afterInstead=False):
        # add a new node with the given newValue immediately BEFORE the node containg targetVal
        # note: this bevahoir can be modified by the bool afterInstead
        # BigO == O(n)

        # walk the list to find it or the end
        found = False
        prev, cur = None, self.head
        while cur is not None:
            if cur.value == targetVal:
                found = True
                break
            prev = cur
            cur = cur.next

        # if found, put it in the chain, as a link right before the node containing value
        if found:
            node = LinkNode(newVal)
            if afterInstead:
                node.next = cur.next
                cur.next = node
            else:
                node.next = cur
                if prev is None:  # edge-case, if the targetVal is first node
                    self.head = node
                else:
                    prev.next = node

        return found

    def insertAfter(self, targetVal: int, newVal: str):
        # add a new node with the given newValue immediately AFTER the node containg targetVal
        # BigO == O(n)
        return self.insertBefore(targetVal, newVal, True)

    def kthFromEnd(self, k):
        # finds the Kth element from the end of the list and returns value for node
        # BigO == O(n)

        # Only positive integers
        if k < 0:
            raise AssertionError(f'WAIT!!! You must pass a positive integer, k:[{k}]')

        ptrA = self.head
        ptrB = self.head

        # Walk ptrA out to "K" elements
        tooSmall = False
        c = 0
        while c < k-1:
            if ptrA.next is None:
                tooSmall = True
                break
            ptrA = ptrA.next
            c += 1
        if tooSmall:
            raise AssertionError(f'WAIT!!! There are not enough elements in the link list for k:[{k}].')

        # Walk ptrA and ptrB out to the end of the list
        # ptrB will point to our requested node
        # note: for short lengths where the value is found before k elements, we are going to skip walking
        # ptrB until the difference between it and ptrA is "k"
        while ptrA.next is not None:
            ptrA = ptrA.next
            if c >= k:
                ptrB = ptrB.next
            c += 1

        return ptrB.value


    def mergeList(self, listA, listB):
        # Merge two lists
        # BigO == O(n)

        ptrA = listA.head
        ptrB = listB.head

        print(f'ptrA:[{ptrA}]')
        print(f'ptrB:[{ptrB}]')
        while ptrA is not None or ptrB is not None:
            print(f'ptrA.value:[{ptrA.value}]')
            print(f'ptrB.value:[{ptrB.value}]')
            if ptrA is not None:
                prev = ptrA
                ptrA = ptrA.next
                prev.next = ptrB
            if ptrB is not None:
                prev = ptrB
                ptrB = ptrB.next
                prev.next = ptrA

        return listA


if __name__ == "__main__":

    listA = LinkList()
    listA.append('apple')
    listA.append('bannana')
    listA.append('orange')

    listB = LinkList()
    listB.append('cheerios')
    listB.append('frosted flakes')
    listB.append('wheaties')

    expected = 'apple,cheerios,bannan,frosted flakes,orange,wheaties'
    listA.mergeList(listA, listB)
    actual = listA.toStr()

    print(f'actual:[{actual}]')

File: data-structures/linked_list/linklist.go
package linklist

import (
	"fmt"
	"os"
	"strings"
)

// LinkNode this is the internal object for individual link-nodes
type LinkNode struct {
	value interface{}
	next  *LinkNode
	prev  *LinkNode
}

// Init constructor for the object
func (node *LinkNode) Init(value interface{}) {
	node.value = value
	node.next = nil
	node.prev = nil
}

// LinkList is the internal data-structure
type LinkList struct {
	head *LinkNode
	// @TODO: comparison_func func
}

// Init constructor
func (list *LinkList) Init( /* @TODO: comparison_func=nil */ ) {
	list.head = nil
	// @TODO: self.comparison_func = comparison_func
}

// toJSON will output the string in JSON format
func (list *LinkList) toJSON() string {
	// @TODO Get a module for this
	return ""
}

// toStr dump list to string
func (list *LinkList) toStr() string {

	// # c,cnt are limiters to make sure we don't go run away
	// # yes, we need cnt=self.count() and not -1, as we walk off list
	c, cnt := 0, list.Count()
	buf := ""
	ptr := list.head
	for ptr != nil {
		s := fmt.Sprintf("%v", ptr.value)
		buf = buf + s + ","
		ptr = ptr.next
		c++
		if c > cnt {
			fmt.Fprintf(os.Stderr, "WAIT!!! Forever Loop!\nRecursive LinkList/Node\nbuf:[{buf}]")
			os.Exit(1)
		}
	}

	if strings.HasSuffix(buf, ",") {
		buf = buf[:len(buf)-1]
	}

	return buf
}

// Insert a value into the link list
func (list *LinkList) Insert(value interface{}) bool {
	node := LinkNode{value: value}
	node.next = list.head
	list.head = &node
	return true
}

// Includes tests to see if a value exists
func (list *LinkList) Includes(value interface{}) bool {
	retVal := false

	cur := list.head
	for cur != nil {
		/*
			if list.comparison_func != nil {
				if list.comparison_func(cur.value, value) {
					retVal = true
					break
				}
			} else {
		*/
		if cur.value == value {
			retVal = true
			break
		}
		/*
			} */
		cur = cur.next
	}

	return retVal
}

// Get will traverse the list and determine if a value exists
func (list *LinkList) Get(value interface{}) interface{} {
	// @TODO: make this return a tuple with error-code
	cur := list.head
	for cur != nil {
		/*
			if list.comparison_func != nil {
				if list.comparison_func(cur.value, value) {
					return cur.value
				}
			} else {
		*/
		if cur.value == value {
			return cur.value
		}
		//		}
		cur = cur.next
	}
	return false
}

// Count the number of items
func (list *LinkList) Count() int {
	cur := list.head

	c := 0
	for cur != nil {
		c++
		cur = cur.next
	}

	return c
}

// Append a value into LinkList
func (list *LinkList) Append(value interface{}) bool {

	// Walk to end of list
	prev, cur := (*LinkNode)(nil), list.head
	for cur != nil {
		prev = cur
		cur = cur.next
	}

	// create the node and add it to the end
	node := LinkNode{value: value}

	if prev == nil {
		list.head = &node
	} else {
		prev.next = &node
	}

	return true
}

// Remove a node from a list, given a specific value
func (list *LinkList) Remove(value interface{}) bool {
	// BigO == O(n*2) ... I could eliminate self.includes(), but I think it's more readable
	// NOTE: only the first one found will be removed

	retVal := false
	if list.Includes(value) {
		//prev, cur = (LinkNode *)(nil), list.head
		prev, cur := (*LinkNode)(nil), list.head

		for cur != nil {

			found := false
			// if list.comparison_func != nil {
			// 	if self.comparison_func(cur.value, value):
			// 		found = True
			// } else  {
			if cur.value == value {
				found = true
			}
			// }

			if found {
				if prev == nil {
					list.head = cur.next
				} else {
					prev.next = cur.next
				}
				retVal = true
				break
			}

			prev = cur
			cur = cur.next
		}

	}

	return retVal
}

// Peek a value from the front of the list
func (list *LinkList) Peek() (bool, interface{}) {
	retVal := interface{}(nil)
	retBool := false
	if list.head != nil {
		retVal = list.head.value
		retBool = true
	}
	return retBool, retVal
}

/*
   def peekHead(self) -> [bool, str]:
       retStr = ''
       retBool = False
       if self.head is not None:
           retStr = self.head.value
           retBool = True
       return [retBool, retStr]


   def insertBefore(self, targetVal: int, newVal: str, afterInstead=False):
       # add a new node with the given newValue immediately BEFORE the node containg targetVal
       # note: this bevahoir can be modified by the bool afterInstead
       # BigO == O(n)

       # walk the list to find it or the end
       found = False
       prev, cur = None, self.head
       while cur is not None:
           if cur.value == targetVal:
               found = True
               break
           prev = cur
           cur = cur.next

       # if found, put it in the chain, as a link right before the node containing value
       if found:
           node = LinkNode(newVal)
           if afterInstead:
               node.next = cur.next
               cur.next = node
           else:
               node.next = cur
               if prev is None:  # edge-case, if the targetVal is first node
                   self.head = node
               else:
                   prev.next = node

       return found

   def insertAfter(self, targetVal: int, newVal: str):
       # add a new node with the given newValue immediately AFTER the node containg targetVal
       # BigO == O(n)
       return self.insertBefore(targetVal, newVal, True)

   def kthFromEnd(self, k):
       # finds the Kth element from the end of the list and returns value for node
       # BigO == O(n)

       # Only positive integers
       if k < 0:
           raise AssertionError(f'WAIT!!! You must pass a positive integer, k:[{k}]')

       ptrA = self.head
       ptrB = self.head

       # Walk ptrA out to "K" elements
       tooSmall = False
       c = 0
       while c < k-1:
           if ptrA.next is None:
               tooSmall = True
               break
           ptrA = ptrA.next
           c += 1
       if tooSmall:
           raise AssertionError(f'WAIT!!! There are not enough elements in the link list for k:[{k}].')

       # Walk ptrA and ptrB out to the end of the list
       # ptrB will point to our requested node
       # note: for short lengths where the value is found before k elements, we are going to skip walking
       # ptrB until the difference between it and ptrA is "k"
       while ptrA.next is not None:
           ptrA = ptrA.next
           if c >= k:
               ptrB = ptrB.next
           c += 1

       return ptrB.value


   def mergeList(self, listA, listB):
       # Merge two lists
       # BigO == O(n)

       ptrA = listA.head
       ptrB = listB.head

       while ptrA is not None or ptrB is not None:
           if ptrA is not None:
               prev = ptrA
               ptrA = ptrA.next
               prev.next = ptrB
           if ptrB is not None:
               prev = ptrB
               ptrB = ptrB.next
               prev.next = ptrA

       return listA
*/
